warp/Network/Wai/Handler/Warp/WithApplication.hs

```diff
@@ -27,26 +32,36 @@ import           Network.Wai.Handler.Warp.Types
 withApplication :: IO Application -> (Port -> IO a) -> IO a
 withApplication = withApplicationSettings defaultSettings
 
+openApplication :: IO Application -> IO (Port, Socket, IO (), Async ())
+openApplication = openApplicationSettings defaultSettings
+
 -- | 'withApplication' with given 'Settings'. This will ignore the port value
 -- set by 'setPort' in 'Settings'.
 --
 -- @since 3.2.7
 withApplicationSettings :: Settings -> IO Application -> (Port -> IO a) -> IO a
-withApplicationSettings settings' mkApp action = do
+withApplicationSettings settings mkApp action =
+  bracket
+    (openApplicationSettings settings mkApp)
+    closeApplication
+    (runApplication action)
+
+openApplicationSettings :: Settings -> IO Application -> IO (Port, Socket, IO (), Async ())
+openApplicationSettings settings' mkApp = do
   app <- mkApp
-  withFreePort $ \ (port, sock) -> do
-    started <- mkWaiter
-    let settings =
-          settings' {
-            settingsBeforeMainLoop
-              = notify started () >> settingsBeforeMainLoop settings'
-          }
-    result <- race
-      (runSettingsSocket settings sock app)
-      (waitFor started >> action port)
-    case result of
-      Left () -> throwIO $ ErrorCall "Unexpected: runSettingsSocket exited"
-      Right x -> return x
+  (port, sock) <- openFreePort
+  started <- mkWaiter
+  let settings =
+        settings' {
+          settingsBeforeMainLoop
+            = notify started () >> settingsBeforeMainLoop settings'
+        }
+  serverAsync <- async (runSettingsSocket settings sock app)
```

Comment:
This will result in a fairly significant change in behavior. The entire `openApplicationSettings` bit is now being run inside a masked exception state, meaning the `runSettingsSocket` call is _also_ in a masked exception state, which is almost certainly not what we want.

Reply:
Excellent catch. `asyncWithUnmask` is what I should have used.